Python/Product/PythonTools/PythonTools/Editor/PythonTextBufferInfo.cs

```diff
@@ -631,46 +633,52 @@ sealed class PythonTextBufferInfo {
         /// </param>
         /// <returns></returns>
         private LineTokenization GetLineTokenization(ITextSnapshotLine line, ref Tokenizer tokenizer) {
+            LineTokenization lineTok;
             int lineNumber = line.LineNumber;
 
-            _tokenCache.EnsureCapacity(line.Snapshot.LineCount);
+            lock (_tokenCache) {
+                _tokenCache.EnsureCapacity(line.Snapshot.LineCount);
 
-            LineTokenization lineTok;
-            int start = _tokenCache.IndexOfPreviousTokenization(lineNumber + 1, 0, out lineTok);
+                int start = _tokenCache.IndexOfPreviousTokenization(lineNumber + 1, 0, out lineTok);
 
-            while (++start <= lineNumber) {
-                var state = lineTok.State;
+                while (++start <= lineNumber) {
+                    var state = lineTok.State;
 
-                if (!_tokenCache.TryGetTokenization(start, out lineTok)) {
-                    if (tokenizer == null) {
-                        tokenizer = MakeTokenizer();
+                    if (!_tokenCache.TryGetTokenization(start, out lineTok)) {
+                        if (tokenizer == null) {
+                            tokenizer = MakeTokenizer();
+                        }
+                        _tokenCache[start] = lineTok = TokenizeLine(tokenizer, line.Snapshot.GetLineFromLineNumber(start), state);
                     }
-                    _tokenCache[start] = lineTok = TokenizeLine(tokenizer, line.Snapshot.GetLineFromLineNumber(start), state);
                 }
             }
 
             return lineTok;
         }
 
-        private void ApplyChangeToTokenCache(ref Tokenizer tokenizer, SnapshotSpan span) {
+        private static void ApplyChangeToTokenCache(TokenCache cache, ref Tokenizer tokenizer, Func<Tokenizer> makeTokenizer, SnapshotSpan span) {
```

Comment:
Not sure I follow the reason for the change - how does making it `static` and passing the instance members it was previously accessing as arguments changes matters, given that it's done under the lock either way?

Reply:
It doesn't, but I feel it makes it a little more obvious that it should be already locked on entry to the function. Maybe it's just me?